test/common/buffer/owned_impl_test.cc

```diff
@@ -1263,6 +1265,120 @@ TEST_F(OwnedImplTest, FrontSlice) {
+  {
+    // Shouldn't trigger slice free yet
+    buffer.drain(4095);
+    EXPECT_EQ(account->balance(), 12288);
+
+    // Trigger slice reclaim.
+    buffer.drain(1);
+    EXPECT_EQ(account->balance(), 8192);
+
+    // Reclaim next slice
+    buffer.drain(std::string("Hello World!").length());
+    EXPECT_EQ(account->balance(), 4096);
+
+    // Reclaim remaining
+    buffer.drain(std::string("Extra Slice").length());
+    EXPECT_EQ(account->balance(), 0);
+  }
+}
+
+TEST(BufferMemoryAccountTest, BufferAccountsForUnownedSliceMovedInto) {
+  std::shared_ptr<BufferMemoryAccount> account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl accounted_buffer;
+  accounted_buffer.bindAccount(account);
+
+  Buffer::OwnedImpl unowned_buffer;
+  unowned_buffer.add("Unaccounted Slice");
+  ASSERT_EQ(account->balance(), 0);
+
+  // Transfer over buffer
+  accounted_buffer.move(unowned_buffer);
+  EXPECT_EQ(account->balance(), 4096);
+
+  accounted_buffer.drain(std::string("Unaccounted Slice").length());
+  EXPECT_EQ(account->balance(), 0);
+}
+
+TEST(BufferMemoryAccountTest, SliceRemainsAttachToOriginalAccountWhenMoved) {
+  std::shared_ptr<BufferMemoryAccount> buffer_one_account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer_one;
+  buffer_one.bindAccount(buffer_one_account);
+  ASSERT_EQ(buffer_one_account->balance(), 0);
+
+  std::shared_ptr<BufferMemoryAccount> buffer_two_account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer_two;
+  buffer_two.bindAccount(buffer_two_account);
+  ASSERT_EQ(buffer_two_account->balance(), 0);
+
+  buffer_one.add("Charged to Account One");
+  EXPECT_EQ(buffer_one_account->balance(), 4096);
+  EXPECT_EQ(buffer_two_account->balance(), 0);
+
+  // Transfer over buffer, still tied to account one.
+  buffer_two.move(buffer_one);
+  EXPECT_EQ(buffer_one_account->balance(), 4096);
+  EXPECT_EQ(buffer_two_account->balance(), 0);
+
+  buffer_two.drain(std::string("Charged to Account One").length());
+  EXPECT_EQ(buffer_one_account->balance(), 0);
+  EXPECT_EQ(buffer_two_account->balance(), 0);
+}
+
+TEST(BufferMemoryAccountTest, SliceRemainsAttachToOriginalAccountWhenMovedUnlessCoalesced) {
+  std::shared_ptr<BufferMemoryAccount> buffer_one_account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer_one;
+  buffer_one.bindAccount(buffer_one_account);
+  ASSERT_EQ(buffer_one_account->balance(), 0);
+
+  std::shared_ptr<BufferMemoryAccount> buffer_two_account = std::make_shared<BufferMemoryAccount>();
+  Buffer::OwnedImpl buffer_two;
+  buffer_two.bindAccount(buffer_two_account);
+  ASSERT_EQ(buffer_two_account->balance(), 0);
+
+  buffer_one.add("Will Coalesce");
+  buffer_two.add("To be Coalesce into:");
+  EXPECT_EQ(buffer_one_account->balance(), 4096);
+  EXPECT_EQ(buffer_two_account->balance(), 4096);
+
+  // Transfer over buffer, slices coalesce, crediting account one.
+  buffer_two.move(buffer_one);
```

Comment:
There's another coalesce case that is worth considering:

```
buffer_three.move(buffer_one);
buffer_three.move(buffer_two);
```

I think that in this case the slice that survives the coalesce originally belonged to buffer_one, so the remaining charge is against account_one regardless of the account associated with buffer_three.

I guess that this is a combination of the consequences of SliceRemainsAttachToOriginalAccountWhenMoved and SliceRemainsAttachToOriginalAccountWhenMovedUnlessCoalesced

Reply:
Added a test case where the two slices coalesce in a third buffer.